moshi/src/main/java/com/squareup/moshi/JsonWriter.java

```diff
@@ -353,12 +353,12 @@ public JsonWriter name(String name) throws IOException {
     if (name == null) {
       throw new NullPointerException("name == null");
     }
-    if (deferredName != null) {
-      throw new IllegalStateException();
-    }
     if (stackSize == 0) {
       throw new IllegalStateException("JsonWriter is closed.");
     }
+    if (deferredName != null) {
+      throw new IllegalStateException();
```

Comment:
Is this what happens when you call `.name("foo").name("bar")`? We should follow up with a good message if so.